x-pack/plugins/security_solution/public/common/containers/matrix_histogram/use_matrix_histogram_async.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { useAsync, withOptionalSignal } from '../../../shared_imports';
+import { DataPublicPluginStart } from '../../../../../../../src/plugins/data/public';
+import { MatrixHistogramQueryProps } from '../../components/matrix_histogram/types';
+import { createFilter } from '../../../common/containers/helpers';
+import {
+  MatrixHistogramQuery,
+  MatrixHistogramRequestOptions,
+  MatrixHistogramStrategyResponse,
+  MatrixHistogramData,
+} from '../../../../common/search_strategy/security_solution';
+import { isErrorResponse, isCompleteResponse } from '../../../../../../../src/plugins/data/common';
+import { getInspectResponse } from '../../../helpers';
+import { InspectResponse } from '../../../types';
+
+export interface GetMatrixHistogramProps extends Omit<MatrixHistogramQueryProps, 'errorMessage'> {
+  data: DataPublicPluginStart;
+  signal: AbortSignal;
+}
+
+export interface GetMatrixHistogramReturnArgs {
+  data: MatrixHistogramData[];
+  inspect: InspectResponse;
+  totalCount: number;
+}
+
+export const getMatrixHistorgram = async ({
+  data,
+  endDate,
+  filterQuery,
+  histogramType,
+  indexNames,
+  stackByField,
+  startDate,
+  threshold,
+  signal,
+}: GetMatrixHistogramProps): Promise<GetMatrixHistogramReturnArgs> => {
+  const request: MatrixHistogramRequestOptions = {
+    defaultIndex: indexNames,
+    factoryQueryType: MatrixHistogramQuery,
+    filterQuery: createFilter(filterQuery),
+    histogramType,
+    timerange: {
+      interval: '12h',
+      from: startDate,
+      to: endDate,
+    },
+    stackByField,
+    threshold,
+  };
+
+  const response = await data.search
+    .search<MatrixHistogramRequestOptions, MatrixHistogramStrategyResponse>(request, {
+      strategy: 'securitySolutionSearchStrategy',
+      abortSignal: signal,
+    })
+    .toPromise();
+
+  if (isCompleteResponse(response)) {
+    return {
+      data: response.matrixHistogramData,
+      inspect: getInspectResponse(response, { dsl: [], response: [] }),
+      totalCount: response.totalCount,
+    };
+  } else if (isErrorResponse(response)) {
+    throw new Error(JSON.stringify(response));
+  } else {
+    return {
+      data: [],
+      inspect: { dsl: [], response: [] },
+      totalCount: 0,
+    };
+  }
+};
+
```

Comment:
Notes, in general is that in the other search strategies I am seeing things like:

```ts
searchSubscription$.unsubscribe
```

But they're missing here along with other parts such as this below?

```ts
notifications.toasts.addWarning(i18n.ERROR_MATRIX_HISTOGRAM);
```

The JSON stringify the response which is not very user friendly in comparison. Is this follow up work to do these parts or just forgotten?

Reply:
I'm working on adding the use of the notifications service, as it does make errors way more readable, thanks for pointing that out. As for the subscriptions, this is to support long-running async searches, which are not yet supported in EQL. We chose to use search strategy because it is something that will be supported so should require some small refactor once that's the case. @rylnd feel free to correct me if any part of that is wrong.

Reply:
I looked it over and you're right, I'm missing a lot of the search strategy functionality. I refactored the code to use the existing hook.

https://github.com/elastic/kibana/pull/78985/commits/b04ec68f197c01ad1e62f0d93edc876b83612aad